### PR TITLE
Natspec4

### PR DESCRIPTION
**Pull request by OdedWx**
Natspec updates to the contracts and interfaces

### CODE REVIEW

**Review by vistra — APPROVED** [review]

File: contracts/Committee.sol [hunk 1]
```diff
@@ -36,6 +37,10 @@ contract Committee is ICommittee, ManagedContract {
 
 	uint8 maxCommitteeSize;
 
+    /// Constructor
+    /// @param _contractRegistry is the contract registry address
+    /// @param _registryAdmin is the registry admin address
+	/// @param _maxCommitteeSize is the maximum number of committee members
```
[the remark this answers is not in the document]
Comment (OdedWx) [reply contracts/Committee.sol]: fixed

File: contracts/Committee.sol [hunk 2]
```diff
@@ -97,26 +139,40 @@ contract Committee is ICommittee, ManagedContract {
...
+	/// @return weights is an array of uint, indicating committee members list weight
+	/// @return certification is an array of bool, indicating the committee members certification status
 	function getCommittee() external override view returns (address[] memory addrs, uint256[] memory weights, bool[] memory certification) {
 		return _getCommittee();
 	}
 
+	/// Returns the currently appointed committee data
+	/// @return generalCommitteeSize is the number of members in the committee
+	/// @return certifiedCommitteeSize is the number of certified members in the committee
+	/// @return totalWeight is the total effective stake / weight of the committee
```
[the remark this answers is not in the document]
Comment (OdedWx) [reply contracts/Committee.sol]: fixed

File: contracts/Committee.sol [hunk 3]
```diff
@@ -170,6 +223,13 @@ contract Committee is ICommittee, ManagedContract {
 	 * Private
 	 */
 
+	/// Checks a member eligibility to join the committee and add if eligible
+	/// @dev Private function called by AddMember and importMembers
```
[the remark this answers is not in the document]
Comment (OdedWx) [reply contracts/Committee.sol]: fixed

File: contracts/Committee.sol [hunk 4]
```diff
@@ -264,6 +337,14 @@ contract Committee is ICommittee, ManagedContract {
 		}
 	}
 
+
```
[the remark this answers is not in the document]
Comment (OdedWx) [reply contracts/Committee.sol:340]: fixed

File: contracts/StakingRewards.sol [hunk 5]
```diff
@@ -545,6 +785,9 @@ contract StakingRewards is IStakingRewards, ManagedContract {
         stakingRewardsState.unclaimedStakingRewards = _stakingRewardsState.unclaimedStakingRewards.sub(total);
     }
 
+    /// Withdraw the tokens allocated for the contract from the staking rewards wallet
```
Comment (ordd) [inline contracts/StakingRewards.sol:788]: typo

File: contracts/StakingRewards.sol [hunk 6]
```diff
@@ -477,6 +697,10 @@ contract StakingRewards is IStakingRewards, ManagedContract {
         emit DelegatorStakingRewardsAssigned(delegator, delegatorStakingRewardsAdded, delegatorStakingRewards.claimed.add(delegatorStakingRewards.balance), guardian, guardianStakingRewards.delegatorRewardsPerToken, delegatorRewardsPerTokenDelta);
     }
 
+    /// Updates a delegator staking rewards state
+    /// @dev query the relevant delegator and committee data from the committee contract amd delegation contract
```
Comment (ordd) [inline contracts/StakingRewards.sol:701]: typo

File: contracts/StakingRewards.sol [hunk 7]
```diff
@@ -344,16 +413,85 @@ contract StakingRewards is IStakingRewards, ManagedContract {
...
 
     // Global state
 
+    /// Returns the annual reward per weight 
+    /// @dev calculates the current annual rewards per weight based on the annual rate and annual cap
     function _getAnnualRewardPerWeight(uint256 totalCommitteeWeight, Settings memory _settings) private pure returns (uint256) {
         return totalCommitteeWeight == 0 ? 0 : Math.min(uint256(_settings.annualRateInPercentMille).mul(TOKEN_BASE).div(PERCENT_MILLIE_BASE), uint256(_settings.annualCap).mul(TOKEN_BASE).div(totalCommitteeWeight));
     }
 
+    /// Calculate the added rewards per weight for the duration based on the committee data
```
Comment (ordd) [inline contracts/StakingRewards.sol:491]: typo

File: contracts/StakingRewards.sol [hunk 8]
```diff
@@ -187,89 +225,91 @@ contract StakingRewards is IStakingRewards, ManagedContract {
...
-        emit EmergencyWithdrawal(msg.sender, address(_token));
-        require(_token.transfer(msg.sender, _token.balanceOf(address(this))), "StakingRewards::emergencyWithdraw - transfer failed");
-    }
+    /*
+    * Governance functions
+    */
 
+    /// Activates staking rewards allocation
+    /// @dev governance function called only by the initialization admin
+    /// @dev On migrations, startTime should be set as the previous contract deactivation time.
```
Comment (ordd) [inline contracts/StakingRewards.sol:311]: reads from the previous contract?

File: contracts/ProtocolWallet.sol [hunk 9]
```diff
@@ -34,20 +40,25 @@ contract ProtocolWallet is IProtocolWallet, WithClaimableMigrationOwnership, Wit
...
-    /// @dev Returns the address of the underlying staked token.
-    /// @return balance IERC20 The address of the token.
+    /// Returns the address of the underlying staked token
+    /// @return balance is the wallet balance
     function getBalance() public override view returns (uint256 balance) {
         return token.balanceOf(address(this));
     }
 
-    /// @dev Transfers the given amount of orbs tokens form the sender to this contract an update the pool.
+    /// Transfers the given amount of orbs tokens form the sender to this contract an updates the pool
```
Comment (ordd) [inline contracts/ProtocolWallet.sol:49]: typo

File: contracts/Protocol.sol [hunk 10]
```diff
@@ -62,6 +81,10 @@ contract Protocol is IProtocol, ManagedContract {
      * Private functions
      */
 
+    /// Check whether a previously set protocol upgrade was executed and returns accordingly the current protocol version
```
Comment (ordd) [inline contracts/Protocol.sol:84]: typo

File: contracts/Lockable.sol [hunk 11]
```diff
@@ -5,22 +5,36 @@ pragma solidity 0.6.12;
...
     bool public locked;
 
+    /// Constructor
+    /// @param _contractRegistry is the contract registry address
+    /// @param _registryAdmin is the registry admin address
     constructor(IContractRegistry _contractRegistry, address _registryAdmin) ContractRegistryAccessor(_contractRegistry, _registryAdmin) public {}
 
+    /// Locks the contract to external non-governance function calls
+    /// @dev governance function called only by the migration manager or an admin
+    /// @dev typically called by the registry contract upon locking all managed contracts
```
Comment (ordd) [inline contracts/Lockable.sol:20]: consider mention "locking all managed contracts" above

File: contracts/Lockable.sol [hunk 12]
```diff
@@ -5,22 +5,36 @@ pragma solidity 0.6.12;
...
+    /// @dev getters and migration functions remain active also for locked contracts
+    /// @dev checked by the onlyWhenActive modifier
     function lock() external override onlyMigrationManager {
         locked = true;
         emit Locked();
     }
 
+    /// Unlocks the contract 
+    /// @dev governance function called only by the migration manager or an admin
+    /// @dev typically called by the registry contract upon unlocking all managed contracts
```
Comment (ordd) [inline contracts/Lockable.sol:30]: consider mention "unlocking all managed contracts" above

File: contracts/GuardiansRegistration.sol [hunk 13]
```diff
@@ -75,34 +103,74 @@ contract GuardiansRegistration is IGuardiansRegistration, ManagedContract {
...
 		orbsAddrs = new address[](guardianAddrs.length);
 		for (uint i = 0; i < guardianAddrs.length; i++) {
 			orbsAddrs[i] = guardians[guardianAddrs[i]].orbsAddr;
 		}
 	}
 
+    /// Returns a guardian's ip
+    /// @dev an unregistered guardian returns 0 ip address
+    /// @param guardian is the guardian to query
+    /// @param ip is the guardian's node ipv4 address as a 32b number
```
Comment (ordd) [inline contracts/GuardiansRegistration.sol:133]: @return is missing

File: contracts/GuardiansRegistration.sol [hunk 14]
```diff
@@ -75,34 +103,74 @@ contract GuardiansRegistration is IGuardiansRegistration, ManagedContract {
...
+    /// @param guardian is the guardian to query
+    /// @param ip is the guardian's node ipv4 address as a 32b number 
 	function getGuardianIp(address guardian) external override view returns (bytes4 ip) {
 		return guardians[guardian].ip;
 	}
 
+    /// Returns the ip of a list of guardians
+    /// @dev an unregistered guardian returns 0 ip address
+    /// @param guardianAddrs is a list of guardians' addresses to query
+    /// @param ips is a list of the guardians' node ipv4 addresses as a 32b numbers
```
Comment (ordd) [inline contracts/GuardiansRegistration.sol:141]: @return is missing

File: contracts/GuardiansRegistration.sol [hunk 15]
```diff
@@ -75,34 +103,74 @@ contract GuardiansRegistration is IGuardiansRegistration, ManagedContract {
...
+    /// @return registered is a bool indicating a guardian address is registered
 	function isRegistered(address guardian) public override view returns (bool) {
 		return guardians[guardian].registrationTime != 0;
 	}
 
+    /// Translates a list guardians Orbs addresses to guardian addresses
+    /// @dev an Orbs address that does not correspond to any registered guardian returns address(0)
+    /// @param orbsAddrs is a list of the guardians' Orbs addresses to query
+    /// @param guardianAddrs is a list of guardians' addresses that matches the Orbs addresses
+	function getGuardianAddresses(address[] calldata orbsAddrs) external override view returns (address[] memory guardianAddrs) {
```
Comment (ordd) [inline contracts/GuardiansRegistration.sol:160]: `@return` is missing

File: contracts/FeesWallet.sol [hunk 16]
```diff
@@ -102,15 +117,22 @@ contract FeesWallet is IFeesWallet, ManagedContract {
         destination.acceptBucketMigration(bucketStartTime, bucketAmount);
     }
 
+    /// Accepts a bucket fees from a old fees wallet as part of a migration
```
Comment (ordd) [inline contracts/FeesWallet.sol:120]: typo

File: contracts/FeesWallet.sol [hunk 17]
```diff
@@ -159,6 +190,7 @@ contract FeesWallet is IFeesWallet, ManagedContract {
         }
     }
 
+    /// Returns the start time of a bucket, sued also to identify the bucket
```
Comment (ordd) [inline contracts/FeesWallet.sol:193]: typo

File: contracts/Subscriptions.sol [hunk 18]
```diff
@@ -162,36 +182,60 @@ contract Subscriptions is ISubscriptions, ManagedContract {
...
+    /// @dev governance function called only by the functional manager
+    /// @param addr is the address of the subscription plan contract
+    function removeSubscriber(address addr) external override onlyFunctionalManager {
+        require(authorizedSubscribers[addr], "given add is not an authorized subscriber");
+
+        authorizedSubscribers[addr] = false;
+        emit SubscriberRemoved(addr);
+    }
+
+    /// Set the genesis reference time delay from the virtual chain creation time
```
Comment (ordd) [inline contracts/Subscriptions.sol:207]: typo

File: contracts/Subscriptions.sol [hunk 19]
```diff
@@ -137,10 +108,59 @@ contract Subscriptions is ISubscriptions, ManagedContract {
...
+    /// @param value is the config record value
+    function setVcConfigRecord(uint256 vcId, string calldata key, string calldata value) external override onlyWhenActive {
+        require(msg.sender == virtualChains[vcId].owner, "only vc owner can set a vc config record");
+        configRecords[vcId][key] = value;
+        emit VcConfigRecordChanged(vcId, key, value);
+    }
+
+    /// Returns the value of a virtual chain config record
+    /// @param vcId is the virtual chain ID
+    /// @param key iis the name of the config record to query
```
Comment (ordd) [inline contracts/Subscriptions.sol:135]: typo

File: contracts/Subscriptions.sol [hunk 20]
```diff
@@ -137,10 +108,59 @@ contract Subscriptions is ISubscriptions, ManagedContract {
...
+    /// @param rate is the virtual chain tier rate as determined by the subscription plan
+    /// @param payer is the address paying for the subscription extension
     function extendSubscription(uint256 vcId, uint256 amount, string calldata tier, uint256 rate, address payer) external override onlySubscriber onlyWhenActive {
         _extendSubscription(vcId, amount, tier, rate, payer);
     }
 
+    /// Sets a virtual chain config record
+    /// @dev may be called only by the virtual chain owner
+    /// @param vcId is the virtual chain ID
+    /// @param key iis the name of the config record to update
```
Comment (ordd) [inline contracts/Subscriptions.sol:125]: typo

File: contracts/Committee.sol [hunk 21]
```diff
@@ -125,10 +181,15 @@ contract Committee is ICommittee, ManagedContract {
 		emit CommitteeSnapshot(addrs, weights, certification);
 	}
 
+
```
Comment (ordd) [inline contracts/Committee.sol:184]: typo

File: contracts/StakingRewards.sol (hunk 5 above)
Comment (OdedWx) [reply contracts/StakingRewards.sol:788]: fixed

File: contracts/StakingRewards.sol (hunk 6 above)
Comment (OdedWx) [reply contracts/StakingRewards.sol:701]: fixed

File: contracts/StakingRewards.sol (hunk 8 above)
Comment (OdedWx) [reply contracts/StakingRewards.sol:311]: May be useful also in non-migration scenarios. For example in order to adjust after a client rate changed

File: contracts/ProtocolWallet.sol (hunk 9 above)
Comment (OdedWx) [reply contracts/ProtocolWallet.sol:49]: fixed

File: contracts/Protocol.sol (hunk 10 above)
Comment (OdedWx) [reply contracts/Protocol.sol:84]: fixed

File: contracts/Lockable.sol (hunk 11 above)
Comment (OdedWx) [reply contracts/Lockable.sol:20]: Tried the have the first @dev comment mention the governance restrictions for all functions that require permissions

File: contracts/Lockable.sol (hunk 12 above)
Comment (OdedWx) [reply contracts/Lockable.sol:30]: Tried the have the first @dev comment mention the governance restrictions for all functions that require permissions

File: contracts/GuardiansRegistration.sol (hunk 13 above)
Comment (OdedWx) [reply contracts/GuardiansRegistration.sol:133]: fixed (param -> return)

File: contracts/GuardiansRegistration.sol (hunk 14 above)
Comment (OdedWx) [reply contracts/GuardiansRegistration.sol:141]: fixed (param -> return)

File: contracts/GuardiansRegistration.sol (hunk 15 above)
Comment (OdedWx) [reply contracts/GuardiansRegistration.sol:160]: fixed (param -> return)

File: contracts/FeesWallet.sol (hunk 16 above)
Comment (OdedWx) [reply contracts/FeesWallet.sol:120]: fixed

File: contracts/FeesWallet.sol (hunk 17 above)
Comment (OdedWx) [reply contracts/FeesWallet.sol:193]: fixed

File: contracts/Subscriptions.sol (hunk 18 above)
Comment (OdedWx) [reply contracts/Subscriptions.sol:207]: fixed

File: contracts/Subscriptions.sol (hunk 19 above)
Comment (OdedWx) [reply contracts/Subscriptions.sol:135]: fixed

File: contracts/Subscriptions.sol (hunk 20 above)
Comment (OdedWx) [reply contracts/Subscriptions.sol:125]: fixed

**Review by 6LYE6M-6MMQ — COMMENTED** [review]
Nuch1991

**Pull request merged**